### PR TITLE
feat(core): Prometheus Counter and Histogram downsampling

### PR DESCRIPTION
**Pull request by vishramachandran**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Capability to downsample Prometheus Counters and Histograms.

Simple strategy:
* Last value of counter or histogram used for every downsample period
* If there is a drop, then the last value before drop and the first value after drop are also included.
* Case of gauges and counters in same schema is handled using downsampler-period-marker which is an abstraction which identifies the row numbers at which downsampling should be done by looking at timestamps of rows or counter values. The period marker is configurable for schemas.

### CODE REVIEW

File: core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  */
+trait DownsamplePeriodMarker {
+
+  def name: PeriodMarkerName
+
+  /**
+    * Ids of Data Columns the marker works on.
+    * The column id values are fed in via downsampling configuration of the dataset
+    */
+  def inputColId: Int
```
Comment (velvia) [inline core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala]: Comment says this reports Ids (plural) but definition only supports one Id.  Make this return Buffer[Int]?

File: core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+  /**
+    * Places row numbers for the given chunkset which marks the
+    * periods to downsample into the result buffer param
+    */
+  def getPeriods(part: ReadablePartition,
```
Comment (velvia) [inline core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala]: Very Java-ish name.  How about just `periods`?

File: core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  }
+  override def name: PeriodMarkerName = PeriodMarkerName.Time
+}
+
+/**
+  * Returns union of the following:
+  * (a) the results from TimeDownsamplePeriodMarker.
+  * (b) row number when counter drops
+  * (c) last row number before counter drops
+  * @param inputColId requires the counter column id
```
Comment (velvia) [inline core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala]: Perhaps add a line or two explaining why we need to add the rows before/after drops

File: core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala [hunk 4]
```diff
@@ -327,7 +327,7 @@ class TimeSeriesShard(val ref: DatasetRef,
     * TSP.ingesting is MUCH faster than bit.get(i) but we need the bitmap for faster operations
     * for all partitions of shard (like ingesting cardinality counting, rollover of time buckets etc).
     */
-  private[memstore] final val activelyIngesting = new EWAHCompressedBitmap
+  private[memstore] final val activelyIngesting = debox.Set.empty[Int]
```
Comment (velvia) [inline core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala]: How is this related to this PR?

File: core/src/main/scala/filodb.core/metadata/Schemas.scala [hunk 5]
```diff
@@ -78,15 +79,18 @@ object DataSchema {
...
 
     for { dataColumns  <- Column.makeColumnsFromNameTypeList(dataColNameTypes)
           downsamplers <- validateDownsamplers(downsamplerNames, downsampleSchema)
+          periodMarker <- validatedDownsamplerPeriodMarker(dowsamplerPeriodMarker)
           valueColID   <- validateValueColumn(dataColumns, valueColumn)
           _            <- validateTimeSeries(dataColumns, Seq(0)) }
     yield {
-      DataSchema(name, dataColumns, downsamplers, Schemas.genHash(dataColumns), valueColID, downsampleSchema)
+      DataSchema(name, dataColumns, downsamplers, Schemas.genHash(dataColumns),
+        valueColID, downsampleSchema, periodMarker)
```
Comment (velvia) [inline core/src/main/scala/filodb.core/metadata/Schemas.scala]: Please line up with name (first param)

File: core/src/main/scala/filodb.core/metadata/Schemas.scala [hunk 6]
```diff
@@ -153,8 +158,13 @@ object PartitionSchema {
 
 /**
  * A Schema combines a PartitionSchema with a DataSchema, forming all the columns of a single ingestion record.
+ *
+ * The downsample member is a var because setting it at construction time can potentially result be impossible
+ * when downsample schema is same as raw schema
+ *
+ * Important Note: Serialization will be tricky since there may be loops in object graph. Avoid if possible.
```
Comment (velvia) [inline core/src/main/scala/filodb.core/metadata/Schemas.scala]: Have to watch for this, I sort of remembered somewhere where Schema needed to be serialized.

File: core/src/main/scala/filodb.core/metadata/Schemas.scala [hunk 7]
```diff
@@ -153,8 +158,13 @@ object PartitionSchema {
...
  */
-final case class Schema(partition: PartitionSchema, data: DataSchema, downsample: Option[Schema]) {
+final case class Schema(partition: PartitionSchema, data: DataSchema, var downsample: Option[Schema] = None) {
```
Comment (velvia) [inline core/src/main/scala/filodb.core/metadata/Schemas.scala]: Can we please find a way to avoid making downsample a var?  Mutability + case classes is a bad mixture, will lead to initialization bugs and related headaches

File: memory/src/main/scala/filodb.memory/format/BinaryVector.scala [hunk 8]
```diff
@@ -176,6 +176,11 @@ trait CounterVectorReader extends VectorDataReader {
    * @param meta CorrectionMeta with total running correction info.  lastValue is ignored
    */
   def updateCorrection(acc: MemoryReader, vector: BinaryVectorPtr, meta: CorrectionMeta): CorrectionMeta
+
+  /**
+    * Identifies row numbers at which histogram or counter has dropped
+    */
+  def dropPositions(acc2: MemoryReader, vector: BinaryVectorPtr): debox.Buffer[Int]
```
Comment (velvia) [inline memory/src/main/scala/filodb.memory/format/BinaryVector.scala]: Can we have the resets function implementation use this method?

File: scripts/schema-truncate.sh [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+cat << EOF
+TRUNCATE ${FILO_ADMIN_KEYSPACE}.checkpoints;
+TRUNCATE ${FILO_KEYSPACE}.${DATASET}_tschunks;
+TRUNCATE ${FILO_KEYSPACE}.${DATASET}_ingestion_time_index;
+EOF
+
+for SHARD in $(seq 0 $((NUM_SHARDS - 1))); do
+
+cat << EOF
```
Comment (velvia) [inline scripts/schema-truncate.sh]: Again, this kind of script is nowhere nearly as convenient as using filo-cli for a whole class of users.  I could not even use this script as is.
Please make this a separate PR, then we can do this properly.
In any case, please update README if desire is for users to use this script, and also update filo-cli instructions.  However, it should really be a separate PR.

File: core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala (hunk 1 above)
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/downsample/DownsamplePeriodMarker.scala]: Only one column in current implementation. I dont want to over-engineer. Easy to change when there is a real use case (multiple counters which can drop at different times for example).

File: core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala (hunk 4 above)
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala]: This was a small omission from previous PR that removed usage of EWAH Bitmap.

File: core/src/main/scala/filodb.core/metadata/Schemas.scala (hunk 6 above)
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/metadata/Schemas.scala]: ok. I'll watch out during testing.

File: core/src/main/scala/filodb.core/metadata/Schemas.scala (hunk 7 above)
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/metadata/Schemas.scala]: Please suggest options. This was the best way out. Immutability and cycles in object graphs is not possible unless I am missing something!

File: memory/src/main/scala/filodb.memory/format/BinaryVector.scala (hunk 8 above)
Comment (vishramachandran) [reply memory/src/main/scala/filodb.memory/format/BinaryVector.scala]: We should eventually use it, outside scope of this PR. Not a simple change because drops across chunks is not covered with this helper.

File: scripts/schema-truncate.sh (hunk 9 above)
Comment (vishramachandran) [reply scripts/schema-truncate.sh]: I added it for convenience since quick capability was missing. We can add other approaches in a different PR.

File: core/src/main/scala/filodb.core/metadata/Schemas.scala (hunk 7 above)
Comment (velvia) [reply core/src/main/scala/filodb.core/metadata/Schemas.scala]: I don't think there is any problem with having a cycle in the case class, except for how you use it ie the `toString` method.

File: memory/src/main/scala/filodb.memory/format/MemoryReader.scala [hunk 10]
```diff
@@ -163,7 +164,7 @@ class OnHeapByteBufferAccessor(buf: ByteBuffer) extends MemoryAccessor {
   }
 
   override def wrapInto(dBuf: DirectBuffer, addr: Long, length: Int): Unit = {
-    dBuf.wrap(buf, addr.toInt, length)
+    dBuf.wrap(buf, origBufPos + addr.toInt, length)
```
Comment (velvia) [inline memory/src/main/scala/filodb.memory/format/MemoryReader.scala]: Good fix.

File: core/src/main/scala/filodb.core/metadata/Schemas.scala (hunk 7 above)
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/metadata/Schemas.scala]: My point is you cant have a cycle with immutable case classes unless you use a var.

**Comment by vishramachandran** [comment]
@velvia Please take another look. The truncate script is currently for dev environments. Since the number of tables have recently increased, the need for this script has become important for dev cycles.
